README.md
# Greenlight fork customised for HT webhub

Is a simple front-end interface for your BigBlueButton server. It provides a minimalistic web-based application that allows users to:

  * Signup/Login with Twitter, Google, or through the application itself.
  * Manage your account settings and user preferences.
  * Create and manage your own personal rooms ([BigBlueButton](https://github.com/bigbluebutton/bigbluebutton) sessions).
  * Invite others to your room using a simple URL.
  * View recordings and share them with others.

It is also completely configurable. This means you can turn on/off features to make it fit your specific use case. For more information on Greenlight and its features, see our [documentation](http://docs.bigbluebutton.org/install/greenlight-v2.html).

## Installation on a BigBlueButton Server

It is designed to work on a [BigBlueButton 2.0](https://github.com/bigbluebutton/bigbluebutton) (or later) server.

For information on installing, checkout our [Installing Greenlight on a BigBlueButton Server](http://docs.bigbluebutton.org/install/greenlight.html#installing-on-a-bigbluebutton-server) documentation.


## Source Code & Contributing

Greenlight is built using Ruby on Rails. Many developers already know Rails well, and we wanted to create both a full front-end to BigBlueButton but also a reference implementation of how to fully leverage the [BigBlueButton API](http://docs.bigbluebutton.org/dev/api.html).

We invite you to build upon Greenlight and help make it better. See [Contributing to BigBlueButton](http://docs.bigbluebutton.org/support/faq.html#contributing-to-bigbluebutton).

We invite your feedback, questions, and suggests about Greenlight too. Please post them to the [developer mailing list](https://groups.google.com/forum/#!forum/bigbluebutton-dev).


changelog.txt
Versione 2.0.4  modifiche dal 21/10/2021 al 20/12/2022 su versione master (greenlight release 2.10.0.1 oct 2021) 

public/strumenti_per_videoconferenze-PROG.jpg
app/views/main/index.html.erb                 (sistemate scritte)
app/views/main/components/_features.html.erb  (rimosse scritte in home page)


Versione 2.0.3  modifiche dal 19/10/2021 al 21/10/2021 su versione master (greenlight release 2.10.0.1 oct 2021) 

app/views/main/components/_features.html.erb (rimosso riferimento a chiamami.online)
config/terms.md            (corretti alcuni testi)


Versione 2.0.2  modifiche dal 18/10/2021 al 19/10/2021 su versione master (greenlight release 2.10.0.1 oct 2021) 
 
 
app/views/user/components/_account.html.erb (commentato il campo "provider", inserita immagine avatar)
app/controllers/application_controller.rb   (impostata "it" come lingua di default)

app/views/layouts/application.html.erb      (aggiunti metatags, aggiunte ottimizzazioni MM, sistemato link ad /info)

app/views/session/new.html.erb             (aggiunta background image)
app/views/session/signin.html.erb          (aggiunta background image)


aggiornati files delle lingue
config/locales/en.yml    (in inglese sostituito bigbluebutton con Riunioni)
config/locales/it.yml    (in italiano sostituito bigbluebutton con Riunioni)



versione 2.0.1 modifiche dal 18/10/2021 al 18/10/2021 su versione master (greenlight release 2.10.0.1)

      create  app/controllers/info_controller.rb
       route  get 'info/termini'
get 'info/privacy'
get 'info/cookies'
get 'info/prezzi'
get 'info/informazioni'
get 'info/help'
get 'info/faq'
get 'info/contact'
      invoke  erb
      create    app/views/info
      create    app/views/info/termini.html.erb
      create    app/views/info/privacy.html.erb
      create    app/views/info/cookies.html.erb
      create    app/views/info/prezzi.html.erb
      create    app/views/info/informazioni.html.erb
      create    app/views/info/help.html.erb
      create    app/views/info/faq.html.erb
      create    app/views/info/contact.html.erb
      invoke  test_unit
      create    test/controllers/info_controller_test.rb
      invoke  helper
      create    app/helpers/info_helper.rb
      invoke    test_unit
      invoke  assets
      invoke    js
      create      app/assets/javascripts/info.js
      invoke    scss
      create      app/assets/stylesheets/info.scss


modifica anche il file  config/route.rb

aggiunta la directory con i testi delle pagine statiche (uno per pagina)  e il file terms
config/info/
config/terms.md

il comando generera anche dei file .erb in (che però vanno poi modificati per mostrare il contenuto)
app/wiews/info (un file erb per ogni pagina)   

il comando modifica
config/route.rb
app/controllers/info_controller.rb

config/routes.rb    (modificati percorsi per le pagine statiche del nuovo controller Info)

modifiche alle immagini

public/favicon.ico                    (sostituita favicon.ico)
public/riunioni_e_lezioni_online.jpg  (aggiunta immagine per metatag open graphs og:image)
public/users/user_placeholder.jpg
public/strumenti_per_videoconferenze.jpg
public/riunioni_e_lezioni_online.jpg

app/assets/images/bbb_logo.png        (sostituita)
app/assets/images/logo_with_text.png  (sostituita)
app/assets/images/favicon.ico  (sostituita)

modifiche per la lingua

config/locales/en.yml    (in inglese sostituito greenlight con Riunioni)
config/locales/it.yml    (file nuovo per lingua italiana)


modifiche alla grafica e ai link di home page e negli include
app/views/main/index.html.erb
app/views/main/components/_features.html.erb (aggiunto paragrafo, modificati box con icone, aggiunto tasto iscrizione)
app/views/shared/_footer.html.erb (rimosso "greenlight", aggiunto paragrafo)
app/views/shared/_header.html.erb (rimosso link ad help "greenlight", sostituito con link alla pagina help)
 


app/controllers/application_controller.rb
# frozen_string_literal: true

# BigBlueButton open source conferencing system - http://www.bigbluebutton.org/.
#
# Copyright (c) 2018 BigBlueButton Inc. and by respective authors (see below).
#
# This program is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free Software
# Foundation; either version 3.0 of the License, or (at your option) any later
# version.
#
# BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.

class ApplicationController < ActionController::Base
  include BbbServer
  include Errors

  before_action :block_unknown_hosts, :redirect_to_https, :set_user_domain, :set_user_settings, :maintenance_mode?,
  :migration_error?, :user_locale, :check_admin_password, :check_user_role

  protect_from_forgery with: :exceptions

  # Retrieves the current user.
  def current_user
    @current_user ||= User.includes(:role, :main_room).find_by(id: session[:user_id])

    if Rails.configuration.loadbalanced_configuration && (@current_user && !@current_user.has_role?(:super_admin) &&
         @current_user.provider != @user_domain)
        @current_user = nil
        session.clear
      end

    @current_user
  end
  helper_method :current_user

  def bbb_server
    @bbb_server ||= Rails.configuration.loadbalanced_configuration ? bbb(@user_domain) : bbb("greenlight")
  end

  # Block unknown hosts to mitigate host header injection attacks
  def block_unknown_hosts
    return if Rails.configuration.hosts.blank?
    raise UnsafeHostError, "#{request.host} is not a safe host" unless Rails.configuration.hosts.include?(request.host)
  end

  # Force SSL
  def redirect_to_https
    if Rails.configuration.loadbalanced_configuration && request.headers["X-Forwarded-Proto"] == "http"
      redirect_to protocol: "https://"
    end
  end

  # Sets the user domain variable
  def set_user_domain
    if Rails.env.test? || !Rails.configuration.loadbalanced_configuration
      @user_domain = "greenlight"
    else
      @user_domain = parse_user_domain(request.host)

      check_provider_exists
    end
  end

  # Sets the settinfs variable
  def set_user_settings
    @settings = Setting.includes(:features).find_or_create_by(provider: @user_domain)
  end

  # Redirects the user to a Maintenance page if turned on
  def maintenance_mode?
    if ENV["MAINTENANCE_MODE"] == "true"
      render "errors/greenlight_error", status: 503, formats: :html,
        locals: {
          status_code: 503,
          message: I18n.t("errors.maintenance.message"),
          help: I18n.t("errors.maintenance.help"),
        }
    end

    maintenance_string = @settings.get_value("Maintenance Banner").presence || Rails.configuration.maintenance_window
    if maintenance_string.present? && cookies[:maintenance_window] != maintenance_string
      flash.now[:maintenance] = maintenance_string
    end
  end

  # Show an information page when migration fails and there is a version error.
  def migration_error?
    render :migration_error, status: 500 unless ENV["DB_MIGRATE_FAILED"].blank?
  end

  # Determines proper locale to be used by calling user_locale with params based on if room owner exists
  def determine_locale(user)
    if user && user.language != 'default'
      user.language
    else
      # GT20211019 Rails.configuration.default_locale.presence || http_accept_language.language_region_compatible_from(I18n.available_locales)
      "it"
    end
  end

  # Sets the appropriate locale.
  def user_locale(user = current_user)
    locale = determine_locale(user)
    begin
      I18n.locale = locale.tr('-', '_') unless locale.nil?
    rescue
      # Default to English if there are any issues in language
      logger.error("Support: User locale is not supported (#{locale}")
      I18n.locale = "en"
    end
  end
  helper_method :user_locale

  # Checks to make sure that the admin has changed his password from the default
  def check_admin_password
    if current_user&.has_role?(:admin) && current_user.email == "admin@example.com" &&
       current_user&.greenlight_account? && current_user&.authenticate(Rails.configuration.admin_password_default)

      flash.now[:alert] = I18n.t("default_admin",
        edit_link: change_password_path(user_uid: current_user.uid)).html_safe
    end
  end

  # Checks if the user is banned and logs him out if he is
  def check_user_role
    if current_user&.has_role? :denied
      session.delete(:user_id)
      redirect_to root_path, flash: { alert: I18n.t("registration.banned.fail") }
    elsif current_user&.has_role? :pending
      session.delete(:user_id)
      redirect_to root_path, flash: { alert: I18n.t("registration.approval.fail") }
    end
  end

  # Relative root helper (when deploying to subdirectory).
  def relative_root
    Rails.configuration.relative_url_root || ""
  end
  helper_method :relative_root

  # Determines if the BigBlueButton endpoint is configured (or set to default).
  def bigbluebutton_endpoint_default?
    return false if Rails.configuration.loadbalanced_configuration
    Rails.configuration.bigbluebutton_endpoint_default == Rails.configuration.bigbluebutton_endpoint
  end
  helper_method :bigbluebutton_endpoint_default?

  def allow_greenlight_accounts?
    return Rails.configuration.allow_user_signup unless Rails.configuration.loadbalanced_configuration
    return false unless @user_domain && !@user_domain.empty? && Rails.configuration.allow_user_signup
    return false if @user_domain == "greenlight"
    # Proceed with retrieving the provider info
    begin
      provider_info = retrieve_provider_info(@user_domain, 'api2', 'getUserGreenlightCredentials')
      provider_info['provider'] == 'greenlight'
    rescue => e
      logger.error "Error in checking if greenlight accounts are allowed: #{e}"
      false
    end
  end
  helper_method :allow_greenlight_accounts?

  # Determine if Greenlight is configured to allow user signups.
  def allow_user_signup?
    Rails.configuration.allow_user_signup
  end
  helper_method :allow_user_signup?

  # Gets all configured omniauth providers.
  def configured_providers
    Rails.configuration.providers.select do |provider|
      Rails.configuration.send("omniauth_#{provider}")
    end
  end
  helper_method :configured_providers

  # Indicates whether users are allowed to share rooms
  def shared_access_allowed
    @settings.get_value("Shared Access") == "true"
  end
  helper_method :shared_access_allowed

  # Indicates whether users should consent recoding when joining rooms
  def recording_consent_required?
    @settings.get_value("Require Recording Consent") == "true"
  end
  helper_method :recording_consent_required?

  # Indicates whether users are allowed to add moderator access codes to rooms
  def moderator_code_allowed?
    @settings.get_value("Room Configuration Moderator Access Codes") == "optional"
  end
  helper_method :moderator_code_allowed?

  # Returns a list of allowed file types
  def allowed_file_types
    Rails.configuration.allowed_file_types
  end
  helper_method :allowed_file_types

  # Allows admins to edit a user's details
  def can_edit_user?(user_to_edit, editting_user)
    return user_to_edit.greenlight_account? if user_to_edit == editting_user

    editting_user.admin_of?(user_to_edit, "can_manage_users")
  end
  helper_method :can_edit_user?

  # Returns the page that the logo redirects to when clicked on
  def home_page
    return admins_path if current_user.has_role? :super_admin
    return current_user.main_room if current_user.role.get_permission("can_create_rooms")
    cant_create_rooms_path
  end
  helper_method :home_page

  # Parses the url for the user domain
  def parse_user_domain(hostname)
    return hostname.split('.').first if Rails.configuration.url_host.empty?
    Rails.configuration.url_host.split(',').each do |url_host|
      return hostname.chomp(url_host).chomp('.') if hostname.include?(url_host)
    end
    ''
  end

  # Include user domain in lograge logs
  def append_info_to_payload(payload)
    super
    payload[:host] = @user_domain
  end

  # Manually handle BigBlueButton errors
  rescue_from BigBlueButton::BigBlueButtonException do |ex|
    logger.error "BigBlueButtonException: #{ex}"
    render "errors/bigbluebutton_error"
  end

  # Manually deal with 401 errors
  rescue_from CanCan::AccessDenied do |_exception|
    if current_user
      render "errors/greenlight_error"
    else
      # Store the current url as a cookie to redirect to after sigining in
      cookies[:return_to] = request.url

      # Get the correct signin path
      path = if allow_greenlight_accounts?
        signin_path
      elsif Rails.configuration.loadbalanced_configuration
        "#{Rails.configuration.relative_url_root}/auth/bn_launcher"
      else
        signin_path
      end

      redirect_to path
    end
  end

  private

  def check_provider_exists
    # Checks to see if the user exists
    begin
      # Check if the session has already checked that the user exists
      # and return true if they did for this domain
      return if session[:provider_exists] == @user_domain

      retrieve_provider_info(@user_domain, 'api2', 'getUserGreenlightCredentials')

      # Add a session variable if the provider exists
      session[:provider_exists] = @user_domain
    rescue => e
      logger.error "Error in retrieve provider info: #{e}"
      @hide_signin = true
      case e.message
      when "No user with that id exists"
        set_default_settings

        render "errors/greenlight_error", locals: { message: I18n.t("errors.not_found.user_not_found.message"),
          help: I18n.t("errors.not_found.user_not_found.help") }
      when "Provider not included."
        set_default_settings

        render "errors/greenlight_error", locals: { message: I18n.t("errors.not_found.user_missing.message"),
          help: I18n.t("errors.not_found.user_missing.help") }
      when "That user has no configured provider."
        if Setting.exists?(provider: @user_domain)
          # Keep the branding
          @settings = Setting.find_by(provider: @user_domain)
        else
          set_default_settings
        end

        render "errors/greenlight_error", locals: { status_code: 501,
          message: I18n.t("errors.no_provider.message"),
          help: I18n.t("errors.no_provider.help") }
      else
        set_default_settings

        render "errors/greenlight_error", locals: { status_code: 500, message: I18n.t("errors.internal.message"),
          help: I18n.t("errors.internal.help"), display_back: true }
      end
    end
  end

  def set_default_settings
    # Use the default site settings
    @user_domain = "greenlight"
    @settings = Setting.find_or_create_by(provider: @user_domain)
  end
end


config/routes.rb
# frozen_string_literal: true

# BigBlueButton open source conferencing system - http://www.bigbluebutton.org/.
#
# Copyright (c) 2018 BigBlueButton Inc. and by respective authors (see below).
#
# This program is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free Software
# Foundation; either version 3.0 of the License, or (at your option) any later
# version.
#
# BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.

Rails.application.routes.draw do

  # info routes.  GT20200328 modificate routes per le pagine info
  # get 'info/termini'
  # get 'info/privacy'
  # get 'info/cookies'
  # get 'info/prezzi'
  # get 'info/informazioni'
  # get 'info/help'
  # get 'info/faq'
  # get 'info/contatti'
  
  get '/termini', to: 'info#termini'
  get '/privacy', to: 'info#privacy'
  get '/cookies', to: 'info#cookies'
  get '/prezzi', to: 'info#prezzi'
  get '/informazioni', to: 'info#informazioni'
  get '/help', to: 'info#help'
  get '/faq', to: 'info#faq'
  get '/contatti', to: 'info#contatti'
  
  get '/health_check', to: 'health_check#all'

  # Error routes.
  match '/401', to: 'errors#unauthorized', via: :all, as: :unauthorized
  match '/404', to: 'errors#not_found', via: :all, as: :not_found
  match '/500', to: 'errors#internal_error', via: :all, as: :internal_error

  # Signin/Signup routes.
  get '/signin', to: 'sessions#signin', as: :signin
  get '/signup', to: 'sessions#new', as: :signup
  post '/signup', to: 'users#create', as: :create_user
  get '/ldap_signin', to: 'sessions#ldap_signin', as: :ldap_signin

  # Redirect to terms page
  match '/terms', to: 'users#terms', via: [:get, :post]

  # Admin resouces
  resources :admins, only: [:index]

  scope '/admins' do
    # Panel Tabs
    get '/rooms', to: 'admins#server_rooms', as: :admin_rooms
    get '/recordings', to: 'admins#server_recordings', as: :admin_recordings
    get '/site_settings', to: 'admins#site_settings', as: :admin_site_settings
    get '/room_configuration', to: 'admins#room_configuration', as: :admin_room_configuration
    get '/roles', to: 'admins#roles', as: :admin_roles
    # Manage Users
    get '/edit/:user_uid', to: 'admins#edit_user', as: :admin_edit_user
    post '/ban/:user_uid', to: 'admins#ban_user', as: :admin_ban
    post '/unban/:user_uid', to: 'admins#unban_user', as: :admin_unban
    post '/invite', to: 'admins#invite', as: :invite_user
    post '/approve/:user_uid', to: 'admins#approve', as: :admin_approve
    get '/reset', to: 'admins#reset', as: :admin_reset
    post '/undelete', to: 'admins#undelete', as: :admin_undelete
    post '/merge/:user_uid', to: 'admins#merge_user', as: :merge_user
    get '/merge_list', to: 'admins#merge_list', as: :merge_list
    # Site Settings
    post '/update_settings', to: 'admins#update_settings', as: :admin_update_settings
    post '/registration_method', to: 'admins#registration_method', as: :admin_change_registration
    post '/coloring', to: 'admins#coloring', as: :admin_coloring
    post '/clear_cache', to: 'admins#clear_cache', as: :admin_clear_cache
    post '/clear_auth', to: 'admins#clear_auth', as: :admin_clear_auth
    post '/log_level', to: 'admins#log_level', as: :admin_log_level
    # Room Configuration
    post '/update_room_configuration', to: 'admins#update_room_configuration', as: :admin_update_room_configuration
    # Roles
    post '/role', to: 'admins#new_role', as: :admin_new_role
    patch 'roles/order', to: 'admins#change_role_order', as: :admin_roles_order
    post '/role/:role_id', to: 'admins#update_role', as: :admin_update_role
    delete 'role/:role_id', to: 'admins#delete_role', as: :admin_delete_role
  end

  scope '/themes' do
    get '/primary', to: 'themes#index', as: :themes_primary
  end

  # Password reset resources.
  resources :password_resets, only: [:new, :create, :edit, :update]

  # Account activation resources
  scope '/account_activations' do
    get '/', to: 'account_activations#show', as: :account_activation
    get '/edit', to: 'account_activations#edit', as: :edit_account_activation
    post '/resend', to: 'account_activations#resend', as: :resend_email
  end

  # User resources.
  scope '/u' do
    # Handles login of greenlight provider accounts.
    post '/login', to: 'sessions#create', as: :create_session

    # Log the user out of the session.
    post '/logout', to: 'sessions#destroy'

    # Account management.
    get '/:user_uid/edit', to: 'users#edit', as: :edit_user
    get '/:user_uid/change_password', to: 'users#change_password', as: :change_password
    get '/:user_uid/delete_account', to: 'users#delete_account', as: :delete_account
    post '/:user_uid/edit', to: 'users#update', as: :update_user
    post '/:user_uid/change_password', to: 'users#update_password', as: :update_password
    delete '/:user_uid', to: 'users#destroy', as: :delete_user

    # All user recordings
    get '/:user_uid/recordings', to: 'users#recordings', as: :get_user_recordings
  end

  # Handles Omniauth authentication.
  match '/auth/:provider/callback', to: 'sessions#omniauth', via: [:get, :post], as: :omniauth_session
  get '/auth/failure', to: 'sessions#omniauth_fail'
  post '/auth/ldap', to: 'sessions#ldap', as: :ldap_callback

  # Users who can't create rooms
  get '/rooms', to: 'rooms#cant_create_rooms', as: :cant_create_rooms

  # Returns a list of users for the shared access list
  get '/shared_access_list', to: 'users#shared_access_list'

  # Room resources.
  resources :rooms, only: [:create, :show, :destroy], param: :room_uid, path: '/'

  # Join a room by UID
  post '/room/join', to: 'rooms#join_specific_room', as: :join_room

  # Extended room routes.
  scope '/:room_uid' do
    post '/', to: 'rooms#join'
    get '/room_settings', to: 'rooms#room_settings'
    post '/update_settings', to: 'rooms#update_settings'
    get '/current_presentation', to: 'rooms#current_presentation'
    post '/preupload_presentation', to: 'rooms#preupload_presentation'
    post '/remove_presentation', to: 'rooms#remove_presentation'
    post '/update_shared_access', to: 'rooms#shared_access', as: :room_shared_access
    delete '/remove_shared_access', to: 'rooms#remove_shared_access', as: :room_remove_shared_access
    get '/shared_users', to: 'rooms#shared_users', as: :room_shared_users
    post '/start', to: 'rooms#start', as: :start_room
    get '/logout', to: 'rooms#logout', as: :logout_room
    post '/login', to: 'rooms#login', as: :login_room
  end

  # Recording operations routes
  scope '/:meetingID' do
    # Manage recordings
    scope '/:record_id' do
      post '/', to: 'recordings#update', as: :update_recording
      patch '/', to: 'recordings#rename', as: :rename_recording
      delete '/', to: 'recordings#delete', as: :delete_recording
    end
  end

  root to: 'main#index'
end
